tokio/src/net/pipe.rs

```diff
@@ -0,0 +1,898 @@
+    /// let mut tx = Sender::open_rw(FIFO_NAME)?;
+    /// // We can asynchronously write to the pipe before any readers.
+    /// tx.write_all(b"hello world").await?;
+    /// # Ok(())
+    /// # }
+    /// ```
+    ///
+    /// # Panics
+    ///
+    /// This function panics if it is not called from within a runtime with
+    /// IO enabled.
+    ///
+    /// The runtime is usually set implicitly when this function is called
+    /// from a future driven by a tokio runtime, otherwise runtime can be set
+    /// explicitly with [`Runtime::enter`](crate::runtime::Runtime::enter) function.
+    pub fn open_rw<P>(path: P) -> io::Result<Sender>
+    where
+        P: AsRef<Path>,
+    {
+        Sender::open_internal(path.as_ref(), true)
+    }
+
+    fn open_internal(path: &Path, read_access: bool) -> io::Result<Sender> {
+        let file = std::fs::OpenOptions::new()
+            .read(read_access)
+            .write(true)
+            .custom_flags(libc::O_NONBLOCK)
+            .open(path)?;
+        let raw_fd = file.into_raw_fd();
+        // Safety: Raw fd was created a valid file.
+        let mio_tx = unsafe { mio_pipe::Sender::from_raw_fd(raw_fd) };
+        Sender::from_mio(mio_tx)
+    }
+
+    fn from_mio(mio_tx: mio_pipe::Sender) -> io::Result<Sender> {
+        let io = PollEvented::new_with_interest(mio_tx, Interest::WRITABLE)?;
+        Ok(Sender { io })
+    }
+
+    /// Creates a new `Sender` from a [`File`].
+    ///
+    /// This function is intended to construct a pipe from a File which represents
+    /// a special FIFO file. The conversion assumes nothing about the underlying
+    /// file; it is left up to the user to open it with writing access and set it
+    /// in non-blocking mode.
+    ///
+    /// You can check first if a [`File`] has the FIFO file type and then use this
+    /// function to get a `Sender`.
+    ///
+    /// # Examples
+    ///
+    /// ```no_run
+    /// use tokio::net::pipe;
+    /// use std::fs::OpenOptions;
+    /// use std::os::unix::fs::{FileTypeExt, OpenOptionsExt};
+    /// # use std::error::Error;
+    ///
+    /// const FIFO_NAME: &str = "path/to/a/fifo";
+    ///
+    /// # async fn dox() -> Result<(), Box<dyn Error>> {
+    /// let file = OpenOptions::new()
+    ///     .write(true)
+    ///     .custom_flags(libc::O_NONBLOCK)
+    ///     .open(FIFO_NAME)?;
+    /// if file.metadata()?.file_type().is_fifo() {
+    ///     let tx = pipe::Sender::from_file(file)?;
+    ///     /* use the Sender */
+    /// }
+    /// # Ok(())
+    /// # }
+    /// ```
+    ///
+    /// # Panics
+    ///
+    /// This function panics if it is not called from within a runtime with
+    /// IO enabled.
+    ///
+    /// The runtime is usually set implicitly when this function is called
+    /// from a future driven by a tokio runtime, otherwise runtime can be set
+    /// explicitly with [`Runtime::enter`](crate::runtime::Runtime::enter) function.
+    pub fn from_file(file: File) -> io::Result<Sender> {
```

Comment:
I suggest that we could add some extra checks on whether it's writeable and actually a pipe fd here, similar to how [`tokio_pipe::PipeFd::from_raw_fd_checked`](https://docs.rs/tokio-pipe/latest/src/tokio_pipe/lib.rs.html#181) performs the check.

Reply:
I'm not really sure that this is necessary.

Reply:
> I'm not really sure that this is necessary.

While it's not absolutely necessary, checking it will rule out some stupid mistakes, e.g. user opened a regular file and passed it to `Sender::from_file`, or they opened the read end of a pipe and passed it here.

It's still runtime error, but better than no error.

Reply:
Also, I think it would be better to use [`OwnedFd`](https://doc.rust-lang.org/std/os/fd/struct.OwnedFd.html) here instead of `File`:

```
fn from_owned_fd(fd: Into<OwnedFd>) -> io::Result<Sender>;
```

since [`impl From<File> for OwnedFd`](https://doc.rust-lang.org/std/os/fd/struct.OwnedFd.html#impl-From%3CFile%3E-for-OwnedFd) makes it possible to create `OwnedFd` from `File`.

There are also `From<ChildStdin>`, `From<ChildStdOut>` and `From<ChildStdErr>` implementation, which would makes this function more useful.

Reply:
> I could totally see people trying to use this to open different things that aren't strictly pipes but are readable or writable only. Do we want to disqualify this use case?

They already have `AsyncFd` for that and `tokio` also provides unix socket, tcp/udp socket, etc.

I don't see why people would want to create `Sender` from `File`, since regular file are blocking, other special files mostly the same.

The only thing that I can think of is people trying to open a named unix socket, but it is already taken care of by [`UnixStream::connect`](https://docs.rs/tokio/latest/tokio/net/struct.UnixStream.html#method.connect), not to mentio that `Sender` does not provide several functions that is available in [`UnixStream`](https://docs.rs/tokio/latest/tokio/net/struct.UnixStream.html#method.connect) and functionalities that will be included (e.g. send fd over unix socket).

Reply:
> I think it would be best to do both File and OwnedFd separately. For now I don't think we can support OwnedFd though due to MSRV.

Oops, then it makes sense to have two separate functions.
It's a shame we can't just have one `from_owned_fd` function.

Reply:
> It's not possible to do those checks independently before calling Sender::open because of [TOCTTOU](https://en.wikipedia.org/wiki/Time-of-check_to_time-of-use).

That's true, it's only possible after the file is opened.

> Maybe this should be more emphasized, or it should be renamed to from_file_unchecked and then we can add a from_file variant with checks.

IMHO renaming it to `from_file_unchecked` and adding a new `from_file` would be better.